### PR TITLE
Add lint-staged and husky for precommit linting.

### PR DESCRIPTION
**Pull request by pksjce**
Dev only change, where while commiting code eslint is run to ensure a clean lint before commit.

Pros - I keep missing this command and have to have a followup lint commit.
Cons - Slows down commit time.

### Screenshots

Screen Shot 2021-12-14 at 1 06 47 pm

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

**Comment by siddharthkp** [comment]
Make robots do the work! love to see it!

I like the idea of running --fix before the commit, bonus points if it creates the commit as well. 

Question: If there is an eslint issue, will this error out and stop you from committing?

**Comment by pksjce** [comment]
Thanks! 

So, you'd want it to fix the lint issue, stage that code and commit everything? I'm just sceptical of giving it so much power. What if it broke a test or something. Though I've never experienced such a drastic lint fix.

Yes, right now, it runs eslint and stops commit if it breaks. We could add prettier too.

**Comment by siddharthkp** [comment]
> So, you'd want it to fix the lint issue, stage that code and commit everything? I'm just sceptical of giving it so much power. What if it broke a test or something. 

Fair enough! It can fix what it can :) 

> We could add prettier too.

Sounds really good!

> Yes, right now, it runs eslint and stops commit if it breaks.

No please please no :( 

We shouldn't stop folks from committing to a feature-branch if eslint/tsc check is failing :( You should be able to get feedback, deploy a preview url, etc without fixing all the rules first.

**Comment by rezrah** [comment]
> No please please no :(
> 
> We shouldn't stop folks from committing to a feature-branch if eslint/tsc check is failing :( You should be able to get > feedback on your code, deploy a preview url, pair with someone etc. without fixing all the rules first.

I kind of agree with this. We also want to also encourage folks to commit small units of code, regularly without punishing them for it.

While you could easily skip husky on those occasions, it'll get pretty tedious adding `--no-verify` each time and it also kind of defeats the point of this change. 

Not a blocker IMO, so I'm happy to go ahead personally. If it becomes a problem, we can circle back to this decision and maybe turn the error into warnings instead if that's preferred?

File: .husky/pre-commit [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npx lint-staged
```
Comment (rezrah) [inline .husky/pre-commit:4]: Should prettier and tsc commands be added too, while we're doing this?

**Review by rezrah — APPROVED** [review]

**Comment by pksjce** [comment]
Yeah, thats mainly what I thought I'll get pushback on. That not commiting part. I'll see what are other approaches

**Review by fariasmateuss — CHANGES_REQUESTED** [review]

**Comment by pksjce** [comment]
@siddharthkp - So this change has the whole precommit hook working when the env variable `HUSKY_ENABLED=true`
A normal git commit will go through as before.
To run the pre-commit husky hook, we'll need `HUSKY_ENABLED=true git commit -m "xoxo"`
For my personal use, I've aliased my git commit for PRC to have this enabled. I think this solves everyone's issues. I get to run the hook and you get to not run the hook by default.

File: package.json [hunk 2]
```diff
@@ -158,5 +161,8 @@
       "webpack": false,
       "running": false
     }
-  ]
+  ],
+  "lint-staged": {
+    "**/*.{js,ts,tsx,md,mdx}": "npm run lint"
```
[the remark this answers is not in the document]
Comment (pksjce) [reply package.json:166]: sure I will add more to this script in the future.

**Pull request merged**